File: 131-palindrome-partitioning/palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> ans=new ArrayList<>();
        List<String> ll=new ArrayList<>();
        pallind(s,ans,ll);
        return ans;

    }
    public void pallind(String s,List<List<String>> ans,List<String> ll){
        if(s.length()==0){
            ans.add(new ArrayList<>(ll));
            return;
        }
        for(int i=1;i<=s.length();i++){
            String t=s.substring(0,i);
            if(check(t)){
                ll.add(t);
                pallind(s.substring(i),ans,ll);
                ll.remove(ll.size()-1);
            }
        }
    }
    public boolean check(String t){
        int i=0;
        int j=t.length()-1;
        while(i<j){
            if(t.charAt(i)!=t.charAt(j)){
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}

File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> ans;
    vector<string> ll;
    vector<vector<string>> partition(string s) {
        solve(s,0);
        return ans;
    }
    void solve(string s,int idx){
        if(idx==s.length()){
            ans.push_back(ll);
            return;
        }
        for(int i=idx;i<=s.length();i++){
            string t=s.substr(idx,i-idx+1);
            if(pall(t)){
                ll.push_back(t);
                solve(s,i+1);
                ll.pop_back();
            }
        }
    }
    bool pall(string s){
        int i=0;
        int j=s.length()-1;
        while(i<j){
            if(s[i]!=s[j]){
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
};